File: k6/load-test.js
/**
 * K6 Load Testing Script for CreatorStudio AI
 * Tests API performance under load
 * 
 * Run: k6 run --out json=results.json load-test.js
 * Run with cloud: k6 cloud load-test.js
 */

import http from 'k6/http';
import { check, sleep, group } from 'k6';
import { Counter, Rate, Trend } from 'k6/metrics';

// Custom metrics for monitoring dashboard
const apiErrors = new Counter('api_errors');
const apiLatency = new Trend('api_latency', true);
const successRate = new Rate('success_rate');

// Test configuration
export const options = {
  // Stages define the load profile
  stages: [
    { duration: '1m', target: 10 },   // Ramp up to 10 users over 1 min
    { duration: '3m', target: 50 },   // Ramp up to 50 users over 3 min
    { duration: '5m', target: 50 },   // Stay at 50 users for 5 min
    { duration: '2m', target: 100 },  // Spike to 100 users
    { duration: '3m', target: 100 },  // Stay at 100 users for 3 min
    { duration: '2m', target: 0 },    // Ramp down to 0 users
  ],
  
  // Thresholds for pass/fail
  thresholds: {
    http_req_duration: ['p(95)<2000'],  // 95% of requests should be under 2s
    http_req_failed: ['rate<0.05'],      // Less than 5% failure rate
    success_rate: ['rate>0.95'],         // 95% success rate
    api_latency: ['p(95)<1500'],         // Custom metric threshold
  },
  
  // Cloud/InfluxDB output for monitoring dashboard
  // Uncomment for cloud execution
  // cloud: {
  //   projectID: 12345,
  //   name: 'CreatorStudio Load Test'
  // }
};

// Base URL from environment or default
const BASE_URL = __ENV.BASE_URL || 'https://legacy-user-fix.preview.emergentagent.com';

// Test user credentials
const TEST_USER = {
  email: 'demo@example.com',
  password: 'Password123!'
};

// Helper function for authenticated requests
function getAuthHeaders(token) {
  return {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${token}`
  };
}

// Helper to record metrics
function recordMetrics(response, endpoint) {
  const success = response.status >= 200 && response.status < 400;
  successRate.add(success);
  apiLatency.add(response.timings.duration);
  
  if (!success) {
    apiErrors.add(1);
    console.log(`ERROR: ${endpoint} - Status: ${response.status}`);
  }
  
  return success;
}

export default function() {
  let authToken = null;

  // ==========================================
  // Authentication Flow
  // ==========================================
  group('Authentication', function() {
    // Health Check
    group('Health Check', function() {
      const healthRes = http.get(`${BASE_URL}/api/health`);
      check(healthRes, {
        'health check status is 200': (r) => r.status === 200,
      });
      recordMetrics(healthRes, '/api/health');
    });

    // Login
    group('Login', function() {
      const loginRes = http.post(
        `${BASE_URL}/api/auth/login`,
        JSON.stringify({
          email: TEST_USER.email,
          password: TEST_USER.password
        }),
        { headers: { 'Content-Type': 'application/json' } }
      );
      
      const loginSuccess = check(loginRes, {
        'login status is 200': (r) => r.status === 200,
        'login returns token': (r) => r.json('token') !== undefined,
      });
      
      recordMetrics(loginRes, '/api/auth/login');
      
      if (loginSuccess) {
        authToken = loginRes.json('token');
      }
    });
    
    sleep(1);
  });

  // Skip remaining tests if auth failed
  if (!authToken) {
    console.log('Authentication failed, skipping authenticated tests');
    return;
  }

  const headers = getAuthHeaders(authToken);

  // ==========================================
  // Dashboard APIs
  // ==========================================
  group('Dashboard APIs', function() {
    // User Profile
    group('Get User Profile', function() {
      const profileRes = http.get(`${BASE_URL}/api/auth/profile`, { headers });
      check(profileRes, {
        'profile status is 200': (r) => r.status === 200,
        'profile has user data': (r) => r.json('name') !== undefined,
      });
      recordMetrics(profileRes, '/api/auth/profile');
    });

    // Credits
    group('Get Credits', function() {
      const creditsRes = http.get(`${BASE_URL}/api/credits`, { headers });
      check(creditsRes, {
        'credits status is 200': (r) => r.status === 200,
        'credits has balance': (r) => r.json('balance') !== undefined,
      });
      recordMetrics(creditsRes, '/api/credits');
    });
    
    sleep(0.5);
  });

  // ==========================================
  // Content Generation APIs
  // ==========================================
  group('Content Generation APIs', function() {
    // Comix AI Styles
    group('Get Comix Styles', function() {
      const stylesRes = http.get(`${BASE_URL}/api/comix/styles`, { headers });
      check(stylesRes, {
        'comix styles status is 200': (r) => r.status === 200,
      });
      recordMetrics(stylesRes, '/api/comix/styles');
    });

    // GIF Maker Styles
    group('Get GIF Styles', function() {
      const gifStylesRes = http.get(`${BASE_URL}/api/gif-maker/styles`, { headers });
      check(gifStylesRes, {
        'gif styles status is 200': (r) => r.status === 200,
      });
      recordMetrics(gifStylesRes, '/api/gif-maker/styles');
    });

    // GenStudio Templates
    group('Get GenStudio Templates', function() {
      const templatesRes = http.get(`${BASE_URL}/api/genstudio/templates`, { headers });
      check(templatesRes, {
        'templates status is 200': (r) => r.status === 200,
        'templates has items': (r) => r.json('templates').length > 0,
      });
      recordMetrics(templatesRes, '/api/genstudio/templates');
    });
    
    sleep(0.5);
  });

  // ==========================================
  // User Manual & Help APIs
  // ==========================================
  group('Help & Documentation', function() {
    // User Manual
    group('Get User Manual', function() {
      const manualRes = http.get(`${BASE_URL}/api/help/manual`, { headers });
      check(manualRes, {
        'manual status is 200': (r) => r.status === 200,
        'manual has features': (r) => r.json('features') !== undefined,
        'manual excludes TwinFinder': (r) => !JSON.stringify(r.json()).includes('twinfinder'),
        'manual includes Comix AI': (r) => r.json('features').comix_ai !== undefined,
      });
      recordMetrics(manualRes, '/api/help/manual');
    });
    
    sleep(0.5);
  });

  // ==========================================
  // Content Vault API
  // ==========================================
  group('Content Vault', function() {
    group('Get Content Vault', function() {
      const vaultRes = http.get(`${BASE_URL}/api/content/vault`, { headers });
      check(vaultRes, {
        'vault status is 200': (r) => r.status === 200,
        'vault has themes': (r) => r.json('themes') !== undefined,
      });
      recordMetrics(vaultRes, '/api/content/vault');
    });
    
    sleep(0.5);
  });

  // ==========================================
  // Analytics API
  // ==========================================
  group('Analytics', function() {
    group('Get User Stats', function() {
      const statsRes = http.get(`${BASE_URL}/api/analytics/user-stats`, { headers });
      check(statsRes, {
        'stats status is 200': (r) => r.status === 200,
      });
      recordMetrics(statsRes, '/api/analytics/user-stats');
    });
    
    sleep(0.5);
  });

  // ==========================================
  // Creator Tools API
  // ==========================================
  group('Creator Tools', function() {
    group('Get Conversion Costs', function() {
      const costsRes = http.get(`${BASE_URL}/api/convert/costs`, { headers });
      if (costsRes.status === 200) {
        check(costsRes, {
          'reel_to_carousel is 10 credits': (r) => r.json('reel_to_carousel') === 10,
          'reel_to_youtube is 10 credits': (r) => r.json('reel_to_youtube') === 10,
          'story_to_reel is 10 credits': (r) => r.json('story_to_reel') === 10,
        });
      }
      recordMetrics(costsRes, '/api/convert/costs');
    });
    
    sleep(0.5);
  });

  // Random think time between iterations
  sleep(Math.random() * 3 + 1);
}

// Teardown function - runs once after all VUs complete
export function teardown(data) {
  console.log('Load test completed');
}

// Handle summary for monitoring integration
export function handleSummary(data) {
  return {
    'results.json': JSON.stringify(data),
    'stdout': textSummary(data, { indent: ' ', enableColors: true }),
  };
}

// Text summary helper
function textSummary(data, options) {
  const indent = options.indent || '';
  let summary = '\n';
  summary += `${indent}=== Load Test Summary ===\n\n`;
  
  // Key metrics
  const metrics = data.metrics;
  summary += `${indent}HTTP Requests:\n`;
  summary += `${indent}  Total: ${metrics.http_reqs.values.count}\n`;
  summary += `${indent}  Rate: ${metrics.http_reqs.values.rate.toFixed(2)}/s\n\n`;
  
  summary += `${indent}Response Times:\n`;
  summary += `${indent}  Avg: ${metrics.http_req_duration.values.avg.toFixed(2)}ms\n`;
  summary += `${indent}  P95: ${metrics.http_req_duration.values['p(95)'].toFixed(2)}ms\n`;
  summary += `${indent}  Max: ${metrics.http_req_duration.values.max.toFixed(2)}ms\n\n`;
  
  summary += `${indent}Errors:\n`;
  summary += `${indent}  Failed: ${metrics.http_req_failed.values.passes} (${(metrics.http_req_failed.values.rate * 100).toFixed(2)}%)\n`;
  
  return summary;
}

File: tests/load/load_test.py
#!/usr/bin/env python3
"""
CreatorStudio AI - Load Testing Script
Medium Load Test: 200 Concurrent Users
Tests all functionalities EXCEPT payment gateway
"""

import asyncio
import aiohttp
import time
import json
import random
import sys
from datetime import datetime
from collections import defaultdict

# Configuration
BASE_URL = sys.argv[1] if len(sys.argv) > 1 else "https://gallery-showcase-43.preview.emergentagent.com"
CONCURRENT_USERS = 200
TEST_DURATION_SECONDS = 120  # 2 minutes
RAMP_UP_SECONDS = 30

# Test credentials
TEST_USERS = [
    {"email": "demo@example.com", "password": "Password123!"},
    {"email": "admin@creatorstudio.ai", "password": "Cr3@t0rStud!o#2026"},
]

# Metrics
metrics = {
    "total_requests": 0,
    "successful_requests": 0,
    "failed_requests": 0,
    "total_logins": 0,
    "successful_logins": 0,
    "response_times": [],
    "errors": defaultdict(int),
    "endpoint_times": defaultdict(list),
}

# Endpoints to test (excluding payment)
ENDPOINTS = [
    # Public endpoints
    ("GET", "/api/health/", False),
    ("GET", "/api/cashfree/products", False),
    ("GET", "/api/pricing/plans?currency=INR", False),
    ("GET", "/api/help/quick-start", False),
    
    # Authenticated endpoints
    ("GET", "/api/auth/me", True),
    ("GET", "/api/wallet/me", True),
    ("GET", "/api/analytics/user-stats", True),
    ("GET", "/api/genstudio/dashboard", True),
    ("GET", "/api/genstudio/templates", True),
    ("GET", "/api/story-series/themes", True),
    ("GET", "/api/story-series/pricing", True),
    ("GET", "/api/story-series/history?limit=5", True),
    ("GET", "/api/challenge-generator/platforms", True),
    ("GET", "/api/challenge-generator/pricing", True),
    ("GET", "/api/tone-switcher/tones", True),
    ("GET", "/api/tone-switcher/pricing", True),
    ("GET", "/api/coloring-book/pricing", True),
    ("GET", "/api/coloring-book/styles", True),
    ("GET", "/api/subscriptions/plans", True),
    ("GET", "/api/privacy/my-data", True),
]


async def login(session, user):
    """Login and get token"""
    try:
        start = time.time()
        async with session.post(
            f"{BASE_URL}/api/auth/login",
            json={"email": user["email"], "password": user["password"]},
            timeout=aiohttp.ClientTimeout(total=10)
        ) as resp:
            elapsed = (time.time() - start) * 1000
            metrics["total_logins"] += 1
            metrics["endpoint_times"]["login"].append(elapsed)
            
            if resp.status == 200:
                data = await resp.json()
                metrics["successful_logins"] += 1
                return data.get("token")
            else:
                metrics["errors"][f"login_{resp.status}"] += 1
                return None
    except Exception as e:
        metrics["errors"][f"login_error"] += 1
        return None


async def make_request(session, method, endpoint, token=None):
    """Make a single request"""
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    
    try:
        start = time.time()
        url = f"{BASE_URL}{endpoint}"
        
        if method == "GET":
            async with session.get(url, headers=headers, timeout=aiohttp.ClientTimeout(total=15)) as resp:
                elapsed = (time.time() - start) * 1000
                metrics["total_requests"] += 1
                metrics["response_times"].append(elapsed)
                metrics["endpoint_times"][endpoint].append(elapsed)
                
                if resp.status in [200, 201, 404]:  # 404 is acceptable for some endpoints
                    metrics["successful_requests"] += 1
                    return True
                else:
                    metrics["failed_requests"] += 1
                    metrics["errors"][f"{endpoint}_{resp.status}"] += 1
                    return False
        
        return True
    except asyncio.TimeoutError:
        metrics["failed_requests"] += 1
        metrics["errors"]["timeout"] += 1
        return False
    except Exception as e:
        metrics["failed_requests"] += 1
        metrics["errors"]["connection_error"] += 1
        return False


async def user_session(session, user_id, start_time):
    """Simulate a single user session"""
    user = random.choice(TEST_USERS)
    
    # Login
    token = await login(session, user)
    if not token:
        return
    
    # Execute random endpoints
    end_time = start_time + TEST_DURATION_SECONDS
    while time.time() < end_time:
        # Pick random endpoint
        method, endpoint, requires_auth = random.choice(ENDPOINTS)
        
        if requires_auth and not token:
            continue
        
        await make_request(session, method, endpoint, token if requires_auth else None)
        
        # Random delay between requests (0.5-2 seconds)
        await asyncio.sleep(random.uniform(0.5, 2))


async def run_load_test():
    """Run the load test"""
    print(f"\n{'='*60}")
    print(f"  CREATORSTUDIO AI - LOAD TEST")
    print(f"  {CONCURRENT_USERS} Concurrent Users | {TEST_DURATION_SECONDS}s Duration")
    print(f"  Target: {BASE_URL}")
    print(f"{'='*60}\n")
    
    start_time = time.time()
    
    # Create connector with connection pooling
    connector = aiohttp.TCPConnector(limit=CONCURRENT_USERS, limit_per_host=CONCURRENT_USERS)
    
    async with aiohttp.ClientSession(connector=connector) as session:
        # Ramp up users
        tasks = []
        users_per_second = CONCURRENT_USERS / RAMP_UP_SECONDS
        
        print(f"[{datetime.now().strftime('%H:%M:%S')}] Starting ramp-up...")
        
        for i in range(CONCURRENT_USERS):
            task = asyncio.create_task(user_session(session, i, start_time))
            tasks.append(task)
            
            # Print progress every 50 users
            if (i + 1) % 50 == 0:
                print(f"[{datetime.now().strftime('%H:%M:%S')}] Spawned {i + 1}/{CONCURRENT_USERS} users")
            
            # Ramp up delay
            await asyncio.sleep(1 / users_per_second)
        
        print(f"[{datetime.now().strftime('%H:%M:%S')}] All {CONCURRENT_USERS} users active. Running test...")
        
        # Wait for all tasks to complete
        await asyncio.gather(*tasks, return_exceptions=True)
    
    total_time = time.time() - start_time
    
    # Calculate statistics
    response_times = metrics["response_times"]
    if response_times:
        avg_response = sum(response_times) / len(response_times)
        sorted_times = sorted(response_times)
        p50 = sorted_times[int(len(sorted_times) * 0.5)]
        p95 = sorted_times[int(len(sorted_times) * 0.95)]
        p99 = sorted_times[int(len(sorted_times) * 0.99)]
    else:
        avg_response = p50 = p95 = p99 = 0
    
    success_rate = (metrics["successful_requests"] / metrics["total_requests"] * 100) if metrics["total_requests"] > 0 else 0
    login_success_rate = (metrics["successful_logins"] / metrics["total_logins"] * 100) if metrics["total_logins"] > 0 else 0
    rps = metrics["total_requests"] / total_time if total_time > 0 else 0
    
    # Print results
    print(f"\n{'='*60}")
    print(f"  LOAD TEST RESULTS")
    print(f"{'='*60}")
    print(f"\n📊 Summary:")
    print(f"   Duration: {total_time:.1f} seconds")
    print(f"   Concurrent Users: {CONCURRENT_USERS}")
    print(f"   Total Requests: {metrics['total_requests']}")
    print(f"   Requests/Second: {rps:.1f}")
    print(f"\n✅ Success Metrics:")
    print(f"   Success Rate: {success_rate:.1f}%")
    print(f"   Successful: {metrics['successful_requests']}")
    print(f"   Failed: {metrics['failed_requests']}")
    print(f"   Login Success Rate: {login_success_rate:.1f}%")
    print(f"\n⏱️ Response Times:")
    print(f"   Average: {avg_response:.0f}ms")
    print(f"   P50: {p50:.0f}ms")
    print(f"   P95: {p95:.0f}ms")
    print(f"   P99: {p99:.0f}ms")
    
    if metrics["errors"]:
        print(f"\n⚠️ Errors:")
        for error, count in sorted(metrics["errors"].items(), key=lambda x: -x[1])[:10]:
            print(f"   {error}: {count}")
    
    # Endpoint performance
    print(f"\n📈 Top 5 Slowest Endpoints:")
    endpoint_avgs = {k: sum(v)/len(v) for k, v in metrics["endpoint_times"].items() if v}
    for endpoint, avg in sorted(endpoint_avgs.items(), key=lambda x: -x[1])[:5]:
        print(f"   {endpoint}: {avg:.0f}ms avg")
    
    print(f"\n{'='*60}")
    
    # Determine pass/fail
    if success_rate >= 90 and p95 < 3000:
        print(f"  ✅ LOAD TEST PASSED")
        print(f"     Success Rate: {success_rate:.1f}% (threshold: 90%)")
        print(f"     P95 Response: {p95:.0f}ms (threshold: 3000ms)")
    else:
        print(f"  ⚠️ LOAD TEST NEEDS ATTENTION")
        if success_rate < 90:
            print(f"     Success Rate: {success_rate:.1f}% (below 90% threshold)")
        if p95 >= 3000:
            print(f"     P95 Response: {p95:.0f}ms (above 3000ms threshold)")
    
    print(f"{'='*60}\n")
    
    # Save results to file
    results = {
        "timestamp": datetime.now().isoformat(),
        "config": {
            "base_url": BASE_URL,
            "concurrent_users": CONCURRENT_USERS,
            "duration_seconds": TEST_DURATION_SECONDS,
        },
        "summary": {
            "total_requests": metrics["total_requests"],
            "successful_requests": metrics["successful_requests"],
            "failed_requests": metrics["failed_requests"],
            "success_rate": success_rate,
            "requests_per_second": rps,
        },
        "response_times": {
            "average_ms": avg_response,
            "p50_ms": p50,
            "p95_ms": p95,
            "p99_ms": p99,
        },
        "logins": {
            "total": metrics["total_logins"],
            "successful": metrics["successful_logins"],
            "success_rate": login_success_rate,
        },
        "errors": dict(metrics["errors"]),
        "passed": success_rate >= 90 and p95 < 3000,
    }
    
    with open("/app/test_reports/load_test_results.json", "w") as f:
        json.dump(results, f, indent=2)
    
    print(f"Results saved to /app/test_reports/load_test_results.json")
    
    return results["passed"]


if __name__ == "__main__":
    success = asyncio.run(run_load_test())
    sys.exit(0 if success else 1)
